<a href="https://colab.research.google.com/github/sulaimanbehzad/Clustering_according_to_item_popularity/blob/main/clustering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clustering of Digikala dataset based on items popularity  


In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder, normalize

### Cloning the github repository in colab

In [44]:
% cd /content/drive/MyDrive/Clustering_according_to_item_popularity/


/content/drive/MyDrive/Clustering_according_to_item_popularity


In [45]:
# ! ls

In [46]:
# ! git clone https://github.com/sulaimanbehzad/Clustering_according_to_item_popularity.git

In [47]:
! git status

On branch main
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Clustering_according_to_item_popularity/

nothing added to commit but untracked files present (use "git add" to track)


In [48]:
! git pull

Already up to date.


In [49]:
# ! git push -u origin main

In [50]:
# ! git commit -m 'redundant files removed'

### Reading and preprocessing the dataset

In [51]:
df = pd.read_excel('/content/drive/MyDrive/Clustering_according_to_item_popularity/keifiat.xlsx')

In [52]:
df.head()

product_id  ...                                      disadvantages
0        3692  ...                                                NaN
1       90213  ...                                          ["ندارد"]
2       59473  ...                                                NaN
3      120499  ...  ["کیفیت پایین اسپیکر\r","حاشیه خالی زیر صفحه ن...
4       67200  ...                                    ["ظرفیت متوسط"]

[5 rows x 12 columns]

In [53]:
df.describe()

product_id       user_id          likes       dislikes
count  100000.000000  1.000000e+05  100000.000000  100000.000000
mean   426482.510140  4.642959e+06       3.114790       1.649460
std    232470.511159  2.619034e+06       8.050031       8.912705
min       300.000000  4.661320e+05       0.000000       0.000000
25%    220871.250000  2.613059e+06       0.000000       0.000000
50%    427858.000000  4.710290e+06       1.000000       0.000000
75%    619397.250000  7.094806e+06       4.000000       1.000000
max    909423.000000  8.623288e+06     854.000000    1344.000000

Select the attributes that will give information about the popularity of the item

In [54]:
df1 = df[['product_id', 'likes', 'dislikes','comment', 'recommend']]

Check the dataframe for null values and handle them

In [55]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   product_id  100000 non-null  int64 
 1   likes       100000 non-null  int64 
 2   dislikes    100000 non-null  int64 
 3   comment     99883 non-null   object
 4   recommend   100000 non-null  object
dtypes: int64(3), object(2)
memory usage: 3.8+ MB


In [56]:
df1.isnull().sum()

product_id      0
likes           0
dislikes        0
comment       117
recommend       0
dtype: int64

looking at the database 'recommend' we can see that \N values also represent null values which wouldn't be considered in the above

In [57]:
df1['recommend'].value_counts()

recommended        36972
\N                 36382
not_recommended    16110
no_idea            10536
Name: recommend, dtype: int64

we can see that beside the 117 missing values we have 36382 records with \N values  
now we understand tha \N too should be considered as missing value


In [ ]:
missing_values = ["\\N", "na", "--", "NA", "n/a"]
df = pd.read_excel('/content/drive/MyDrive/Clustering_according_to_item_popularity/keifiat.xlsx', na_values = missing_values)
df1 = df[['product_id', 'likes', 'dislikes','comment', 'recommend']]


after importing database with newly found aspects on NA we are going to handle null values

In [ ]:
df1.isnull().sum()

In [ ]:
df1['recommend'].value_counts()

In [ ]:
most_common = df1["recommend"].value_counts().idxmax
# df1['recommend'].fillna(median, inplace=True)
most_common

replacing the null values with one of the above would falsely impact the clustering so let's take another approach:  
1. use interpolation to find values
2. fill with mean/median

in order to use interpolation let us first use label encoder to obtain numerical values

In [ ]:
df1['recommend'].fillna(value='null_values', inplace=True)
df1.isnull().sum()

In [ ]:
le = LabelEncoder()
le.fit(df1['recommend'])
le.classes_

In [ ]:
recomm_trans = le.transform(df1['recommend'])
recomm_trans

In [ ]:
df1['recommend'] = recomm_trans

In [ ]:
df1[['recommend']]

In [ ]:
df1['recommend'] = df['recommend'].replace(2,np.nan)
df1['recommend']

In [ ]:
df_filled = df1.interpolate(method='linear', axis=1)

In [ ]:
df2